FrostCord/FrostCord
# Friend System

# Friend System
Adds a system allowing users to become friends on the app.

# API Changes
- Adds `friends.service.ts` which provides functions to get the user's friends/friend requests + cast them to detailed relations
- Adds a friend screen to view current friends and pending friend requests
- Adds `@radix/react-context-menu` and `@radix/react-dropdown-menu` for easy on-click context menus and right-click conext menus
- Adds the ability to right click users's profiles/username (in the message header bar)
- Moves the online presence realtime channel to use a store and an object to access users by id instead of hackily hooking into `presenceState()`

# Testing:
- Verify you cannot send a friend request to yourself
- Verify sending a request to another user does the following:
  - **AS THE REQUESTING USER**
    - Replaces the "send friend request" button with a "cancel friend request" button in the context menu
    - Adds the request to the `pending` tab in the friends list, with only an `X` to cancel the request being present
  - **AS THE RECEPIENT**
    - Verify that the context menu for the user who sent the friend request provides options to accept or reject the friend request, and that these both function as expected
    - Verify that in the friends menu, the pending request has two buttons, a checkmark to accept, and an X to reject the request, and that these both function as expected
- Verify that after accepting a friend request, the right click context menu for either user provides the ability to remove the user from friends.
